### PR TITLE
Add concat banner/footer try-catch blocks

### PR DESCRIPTION
**Pull request by dtothefp**
This PR addresses https://optimizely.atlassian.net/browse/MKT-862.  The goal of the PR is to wrap the global JS bundle and page JS for staging and production in try-catch blocks in order to report bundle and page JS errors to Google Analytics.  This was accomplished by adjusting the concat banner/footer strings in the `grunt config` task and by adding a `targetName` variable inside each JS bundle by using the `process` option in the `grunt concat` task.  

The `targetName` variable is set equal to the current concat target.  Because each bundle is wrapped within a closure the `targetName` variable will be scoped to each bundle and therefore not overwritten by additional JS bundles on the page.  The `targetName` variable is later used within the `catch` block to report to GA whether a JS error occurred within the global JS bundle, or within a page/layout bundle.  *Note: if an error occurs in the global JS bundle it is likely an error will occur as well in a page/layout level bundle as it most likely uses JS from the global bundle.*

Due to the fact that this PR addresses primarily code that is used in staging/production, it is difficult to test locally.  The best way to test locally is to adjust the concat banner and footer strings in the `grunt config:dev` task/target.  Therefore, to test locally alter the `grunt config:dev` target as follows to include the try-catch block as well as a debugger statement:

```js
concat_banner: '(function($, w, d){ \n\n' +
                       '  window.optly = window.optly || {}; \n\n' +
                       '  window.optly.mrkt = window.optly.mrkt || {}; \n\n' +
                       '  window.linkPath = "/dist"; \n\n' + 
                       '  try { \n\n',
        concat_footer: '  } catch(error){ \n\n' +
                       '    var path = window.location.pathname;\n\n' +
                       '    var trimpath = path.lastIndexOf("/") === path.length - 1 ? path.substr(0, path.lastIndexOf("/")) : path;\n\n' + 
                       '    debugger;window.ga("send", "event", targetName + " JavaScript Error", trimpath, error);\n\n' + 
                       '  }\n\n' +
                       '})(jQuery, window, document);'
```

Next, introduce some sort of error in the global JS bundle and/or page level bundle.  Now run `grunt server` and make sure to open the JS console in the chrome inspector.  JS execution should halt at the breakpoint inside the `catch` block and the `targetName` variable can be inspected as well as the `error` object.

### CODE REVIEW

**Comment by dtothefp** [comment]
retest this please

**Comment by dtothefp** [comment]
@kylerush I've successfully introduced an error into staging and it build correctly 

http://optimizely-marketing-site-staging.s3-website-us-east-1.amazonaws.com/dfoxpowell/catcherror-concat-bundles/

The only problem is that sometimes the window.ga object is undefined, I'm assuming this has to do with asynchronus loading.  Should I be somehow pushing into the `ga` queue or using some sort of `ready` function?

**Comment by dtothefp** [comment]
retest this please

**Comment by dtothefp** [comment]
retest this please

**Comment by dtothefp** [comment]
retest this please

**Comment by dtothefp** [comment]
@kylerush Currently I'm manually throwing an error within both the global and page bundles.  It is hardcoded in the concat banner strictly for testing in staging.  When we want to deploy we will remove the error and merge.  

Changed the concat footer to report to GA through Segment:

```js
w.analytics.ready(function() { 
  w.analytics.track(trimpath + ": " + targetName, 
    {
      category: "JavaScript Error", 
      label: error
    }, 
    { 
      integrations: {
        "All": false, 
        "Google Analytics": true
      } 
  }); 
});
```

Values submitted to segment for bundle.js errors:
screen shot 2015-01-09 at 3 38 00 pm

Values submitted to segment for page/layout.js errors:
screen shot 2015-01-09 at 3 38 35 pm

**Comment by dtothefp** [comment]
retest this please

File: grunt/concat.js [hunk 1]
```diff
@@ -1,60 +1,73 @@
...
+
+  return split.splice(split.length - 2).join('/');
+};
+module.exports = function(grunt, options){
+  return {
+    namespacePages: {
+      options: {
+        banner: '<%= grunt.config.get("concat_banner") %>',
+        footer: '<%= grunt.config.get("concat_footer") %>',
+        process: function(src, filepath) {
```
Comment (stewsmith) [inline grunt/concat.js]: Where is this function getting called?

File: grunt/config.js [hunk 2]
```diff
@@ -33,8 +33,11 @@ var config = {
                        '  window.linkPath = "" \n\n' +
                        '  try { \n\n',
         concat_footer: '  } catch(error){ \n\n' +
-                       '  //report errors to GA \n\n' +
-                       '  window.console.log("js error: " + error);' +
+                       '    if(typeof error === "object") { error = JSON.stringify(error) }; \n\n' +
+                       '    if (typeof targetName === "undefined") { var targetName = "namespaceGlobal" }; \n\n' +
```
Comment (stewsmith) [inline grunt/config.js]: This might be a good candidate for `!!targetName` in case `targetNam

File: grunt/config.js [hunk 3]
```diff
@@ -54,11 +57,14 @@ var config = {
...
                        '  window.optly.mrkt = window.optly.mrkt || {}; \n\n' +
-                       '  window.linkPath = "<%= gitinfo.local.branch.current.name %>" \n\n' +
+                       '  window.linkPath = "<%= gitinfo.local.branch.current.name %>"; \n\n' +
                        '  try { \n\n',
-        concat_footer: '  } catch(error){ \n\n' +
-                       '  //report errors to GA \n\n' +
-                       '  window.console.log("js error: " + error);' +
+        concat_footer: '  ;throw new Error("Big Fat Error!!!") } catch(error){ \n\n' +
+                       '    if(typeof error === "object") { error = JSON.stringify(error) }; \n\n' +
+                       '    if (typeof targetName === "undefined") { var targetName = "namespaceGlobal" }; \n\n' +
```
Comment (stewsmith) [inline grunt/config.js]: Same comment as above.

File: website-guts/assets/js/components/modals.js [hunk 4]
```diff
@@ -125,7 +125,7 @@ window.optly.mrkt.modal.open = function(modalArgs) {
   if(modalState.type !== undefined) {
     window.optly.mrkt.modal.close({ modalType: modalState.type, track: false });
   }
-
+
```
Comment (stewsmith) [inline website-guts/assets/js/components/modals.js]: Could you remove this whitespace? :neckbeard:

**Comment by stewsmith** [comment]
I've tested this on staging and it caught the error as expected.

LGTM once my small comments are addressed. :+1:

**Pull request merged**